solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,429 @@
+
+				for (int i = 0; i < fieldMapList.size(); i++) {
+					Map<String, String> fieldMap = fieldMapList.get(i);
+
+					columnNames.add(fieldMap.get("name"));
+					columnTypes.add(ColumnType.VARCHAR);
+				}
+
+				String[] columnNamesArr = new String[columnNames.size()];
+				columnNamesArr = columnNames.toArray(columnNamesArr);
+
+				ColumnType[] columnTypesArr = new ColumnType[columnTypes.size()];
+				columnTypesArr = columnTypes.toArray(columnTypesArr);
+
+				tableDef = new SimpleTableDef(indexName, columnNamesArr,
+						columnTypesArr);
+			}
+		} catch (IOException e) {
+			logger.error("Failed to parse schema", "", e);
+		}
+
+		return tableDef;
+	}
+
+	@Override
+	protected Schema getMainSchema() throws MetaModelException {
+		try {
+			final MutableSchema theSchema = new MutableSchema(
+					getMainSchemaName());
+
+			final MutableTable table = tableDef.toTable().setSchema(theSchema);
+			final Column idColumn = table.getColumnByName(FIELD_ID);
+			if (idColumn != null && idColumn instanceof MutableColumn) {
+				final MutableColumn mutableColumn = (MutableColumn) idColumn;
+				mutableColumn.setPrimaryKey(true);
+			}
+			theSchema.addTable(table);
+
+			return theSchema;
+		} catch (Exception e) {
+			throw new MetaModelException("Schema retrieval failed " + e);
+		}
+	}
+
+	@Override
+	protected Number executeCountQuery(Table table,
+			List<FilterItem> whereItems, boolean functionApproximationAllowed) {
+		String queryStr = "*:*";
+
+		if (whereItems != null) {
+			for (int i = 0; i < whereItems.size(); i++) {
+				FilterItem whereItem = whereItems.get(i);
+				queryStr = whereItem.toString();
+			}
+		}
+
+		QueryResponse response = selectRows(table, null, queryStr, 0, 0);
+		return response.getResults().getNumFound();
+	}
+
+	private String processSingleFilterItem(FilterItem filterItem) {
+		String clause = "";
+
+		if (filterItem != null) {
+			SelectItem item = filterItem.getSelectItem();
+			OperatorType operator = filterItem.getOperator();
+			Object operand = filterItem.getOperand();
+
+			if (item == null || operator == null || operand == null) {
+				clause = " " + filterItem.getExpression();
+			} else {
+				String columnName = item.getColumn().getName();
+				clause = " " + columnName + operator.toSql() + operand;
+			}
+		}
+
+		return clause;
+	}
+
+	private String createSearchClause(FilterItem filterItem, String clause) {
+		FilterItem[] filterItems = filterItem.getChildItems();
+
+		if (filterItems != null && filterItems.length > 0) {
+			for (FilterItem filterItemInner : filterItems) {
+				if (filterItemInner != null) {
+					clause += " " + processSingleFilterItem(filterItemInner);
+				}
+
+				return createSearchClause(filterItemInner, clause);
+			}
+		}
+
+		if (filterItem != null) {
+			clause += " " + processSingleFilterItem(filterItem);
+		}
+
+		return clause;
+	}
+
+	@Override
+	public DataSet executeQuery(Query q) throws MetaModelException {
```

Comment:
I would not override this method for our initial attempt at doing Solr ... Waaaay too much complexity involved in this method, so my advice is to start with just materializeMainTable and make that work well.